sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala

```diff
@@ -79,6 +79,38 @@ class IncrementalExecution(
       StreamingTransformWithStateStrategy :: Nil
   }
 
+  // Methods to enable the use of AsyncLogPurge
+  protected val minLogEntriesToMaintain: Int =
+    sparkSession.sessionState.conf.minBatchesToRetain
+
+  val errorNotifier: ErrorNotifier = new ErrorNotifier()
+
+  override protected def purge(threshold: Long): Unit = {}
+
+  override protected def purgeOldest(planWithStateOpId: SparkPlan): Unit = {
+    planWithStateOpId.collect {
+      case ssw: StateStoreWriter =>
+        ssw.operatorStateMetadataVersion match {
+          case 2 =>
+            val fileManager = new OperatorStateMetadataV2FileManager(
+              new Path(checkpointLocation, ssw.getStateInfo.operatorId.toString),
+              ssw.stateSchemaFilePath(Some(StateStoreId.DEFAULT_STORE_NAME)),
+              hadoopConf)
+              fileManager.keepNEntries(minLogEntriesToMaintain)
```

Comment:
The logic here might not be very straight forward. We need to keep the metadata for the earliest retained state basically

Reply:
Sorry, what does this mean?
If we want to keep the metadata for the latest N runs, we can just sort the files by batchId, right?